### ISSUE TITLE
Handling of secrets, systemd's role in it

### ISSUE THREAD

**Issue by mutech** [systemd/systemd#26208]
### Component

systemd

### Is your feature request related to a problem? Please describe

# The Problem

I want to avoid storing secrets on the file system or in places with an equivalent exposure.

FS permissions provide a decent protection for information, but always only work at the level of granularity of a user id. If an attacker gains access to a process running as that user, they generally also have access to secrets exposed to that user.

In a scenario where a service is running as a system user, it typically needs a secret only in the context of that service. It should be sufficient if only the processes controlled/started by systemd would have access to such secrets, not every process running as that user or worse every process having access to the file in which such a secret is stored.

There are many techniques to solve this problem, but very few of them actually work and it's an incredible effort to do it and get it right.

As a consequence of that difficulty, the industry seems to have concluded that this is a problem to be solved by cloud providers and thus most docker images get their secrets from environment variables or expect to find them in some variation of `credentials.ini`. While this is a reasonable conclusion, I keep getting mails informing me that my account has been found in some pool dripping out of some database leak.

### Describe the solution you'd like

I imagine the following procedure to solve this issue:

I am assuming there is or will be a mechanism implemented by the kernel, that provides the systemd process (pid 1) exclusive read access to secrets identified by a pair of UID and key and a policy spec for the usage of that secret (see below).

A user can provide systemd with a secret and specify how it can later retrieve it. The user can specify if they can later delete that secret (otherwise that would require root). The user can specify which service can use this secret, but access to that secret can never be extended from that point forward, only restricted or deleted (if so specified).

A service specification would then reference the secret in a service definition and systemd provides the secret in a secure way ensuring that a service process gets access to that secret once it's started. As long as this service is running, systemd will refuse to provide that secret to anybody. If the service needs that information again, it can exit and systemd will restart it.

This could be extended to charge systemd with a verification of the integrity of that service, f.e. by verifying signatures.

If the user does not control the secret, it's up to root to associate the secret with a server. That would then be the equivalent of the situation after the user provided the secret to systemd (when they can no longer extend the exposure of the secret).

### The benefit of this approach

If the kernel does not provide a means to exclusively pass secrets to init now, there is at least the benefit that there would be a design pattern that people could start to follow instead of using a wide variety of bad practices to handle secrets. Once such a mechanism is available, the whole ecosystem would benefit from the additional protection of secrets.

Even if there is no immediate protection, alone the fact that such secrets are moved out of the way could improve security. Backup routines would not collect secrets and store them somewhere, where an operator or procedures handling this data down the way are not aware of what is stored there. Docker images like certbot would probably not be designed to require a `credential.ini` file exposing a DNS api token probably in /etc/letsencrypt that might well be world readable.

systemd is in the unique position to have pid 1 and with great power...

**Comment by mutech** [comment]
To clarify that: The in my opinion most important feature would be to allow services to request a secret by UID+key, abstracting away the location of the storage from operations. Even if that would just move the location from whatever the operator otherwise would choose to the discretion of systemd, it would create a pattern that replaces chaos with something that can be fixed later. It would also shift the responsibility from us to systemd which naturally is a good thing! :-)

The implementation of restricted access to that secret is something I only have a vague idea how to implement and that needs to be done right. This would require at least a crypto hash of a service definition to which the stored secret is assigned and that has the risk that the user changing the service definition is not aware of the consequences leading to the service to fail. I think that's worth the price, but someone reading logs at a Friday afternoon might see that differently.

**Comment by poettering** [comment]
Are you aware of systemd's "credentials" concept? → https://systemd.io/CREDENTIALS

**Comment by mutech** [comment]
> Are you aware of systemd's "credentials" concept? → https://systemd.io/CREDENTIALS

Yes, but I might have misunderstood it. What I understood is this:

The actual credentials are stored in unit files (SetCredential) or in the file system (LoadCredential with regular files) or handled by a third party (with domain sockets).

If a third party (a process providing credentials via a socket) is involved, this is out of scope (it's that party handling the issue then).

When the user provides the secret, the result is no different from just storing it in a file, because that's where systemd is reading it from.

I'm not an expert in how systemd does things, far from it, so I hope I'm not wasting your time.

Here is how I came across this issue in its recent incarnation:

I set up a couple of services running in containers. I choose docker-rootless as a simple platform. Rootless so that a vulnerable container has a harder time getting root on the host. The default setup for rootless from Docker requires the service user to have a self-writable home directory, where among other things the user-service is specified.

I did not find a way using systemd credentials, that helped me isolate secrets from a compromised container. Whether the secret is stored in a unit file somewhere below `~/.local` or in `~/credentials.ini` does not really make so much of a difference, even if the unit file only references another path.

What I am looking for is an institution that takes custody of the secret and ensures that only the valid recipient has access to it.

In a real live analogy, systemd should have the role of a notary but it is actually more like the neighbour being told to find the keys in the flower pot while we are out in order to feed the cat and water the plants.

Of course systemd can't do magic. If it had to keep secrets across reboots, it has to store them somewhere. If there is support from the outside, be that hardware, a hypervisor or a domain socket, then this does the job just fine. But it should not be the role of a user or service to setup such an information channel (the unix domain socket), that should be the responsibility of the system (kernel/systemd/infrastructure) and there should be a plug where a service can connect to.

Systemd became so popular, that I think it's perfectly reasonable to just assume it's there. A lot tools, such as docker, already do that.

It would be good enough if the systemd credentials mechanism would provide another variable `SetCredentialKey` and provide a means for root to provide credentials for a UID matching the key. Root's involvement is aimed at preventing a compromised process with the service UID from just asking for the secret.

Did I get it right? Do you see my issue?

**Comment by mutech** [comment]
Btw., I am aware that I can do everything I'm asking systemd to do myself (in my role as operator). But people expect me to get done with the job, because I promised them that docker is simple and getting this and that service up and running is no big deal either. They also expect me to do it right in terms of security, but they only say that and don't really care. That would of course change right that instant where something bad happens.

I could do my job and just put the api token for their DNS in a unit file and I guess nobody would really blame me, because that's what everybody in this segment of the industry does (I'm exaggerating a bit), but this somehow feels unethical.

But when I try to get it right, this is actually a personal hobby. I can't justify to spend more time finding a way to handle secrets correctly than it took to setup the whole service. This is all very specific to my circumstances and my level of expertise, but looking at the state of information security, the problem is just symptomatic for the fact that the wrong people take care of security issues.  It shouldn't be an ops guy designing reliable information channels.

Someone in the documentation you quoted explains that `SetCredential` is not the right tool to use for sensible information. That's all good and dandy, but what then is the right tool? Using `LoadCredential` pointing to a unix socket is one answer, but that only translates to "ask the guy who knows better and waits at the other end of the domain socket". In my case, there is no socket.

**Comment by poettering** [comment]
I am not sure I grok the problem. Where exactly do you want the secret to be stored? on disk? only accessible by root? Then LoadCredential= is what you could use: it can copy secrets files from a root-only file into a per-service file that is only accessible to the user associated with that service (i.e. what you configure via User=).

or you can use LoadEncryptedCredential= which goes further and authenticates/decrypts via TPM2.

what else do you need?

**Comment by mutech** [comment]
> Where exactly do you want the secret to be stored? on disk?

That should be none of the concern of a service. This could be something systemd implements internally, delegates to a plugin or is up to the system operator (as opposed to a service operator).

> only accessible by root?

From the service perspective, the secret should only be accessible to the instance of the service process that is managed by systemd.

How the secret is protected at rest and who does the protecting is - again from the service perspective - an implementation detail.

> Then LoadCredential= is what you could use: it can copy secrets files from a root-only file into a per-service file that is only accessible to the user associated with that service (i.e. what you configure via User=).

Correct me if I'm wrong, but the place from where `LoadCredential` copies the secret needs to hold that secret persistently, doesn't it? After a reboot, systemd will have to go get the secret from that place again. If it's a socket, then the process feeding the secret into the socket will have to get it from some place (file).

If I have to manage this secret anyway, why would I need to pass it through systemd at all?

What I'm looking for is kind of an SOP where a service can rely on a mechanism to securely store secrets and in this case benefit from the special role of systemd being the unique entity controlling all services and user sessions.

The value that systemd adds and what makes it the obvious choice for a keeper of secrets is:
* It's the entity starting and in control of service processes, the one best qualified to determine if a process is a service instance and also the one that has a communication channel open to that process.
* It's also a privileged process that has to be trusted and has the special protection/role of pid 1. It's hard(er) to impersonate or tamper with.

The question whether the persistent storage used is actually more secure than putting it in a file protected by unix permissions is an implementation detail, albeit an important one. I'm actually looking for the separation of concerns. Confidentiality should be a core design element. Currently it's an afterthought that everybody handles according to their personal preferences (no critique intended, just describing the situation as I see it).

Wasn't the whole point of systemd to use an appropriate abstraction for what init was no longer fit to implement? As I understand it, systemd is an abstraction layer that is very similar to the kernel. Systemd does deals in terms of sessions (services/units/states) while the kernel handles processes (resources). I think secrets should be for systemd what capabilities are for the kernel, something that's tightly controlled and handled carefully.

Concretely, there are two things I need:

Define a mechanism how a service user (as opposed to a service instance - e.g. a running process) can hand over a secret to systemd, that is from that moment on only available to valid instance of a service (a process owned by the service user that was started by systemd or is otherwise reliably identifiable as a valid service instance).

The secret can then be revoked or updated by the service user, but not read. That's what I meant by systemd taking custody of the secret.

**Comment by mutech** [comment]
Risking to go off-topic in my own ticket, I'm elaborating a bit to explain my original concern.

I'm currently in the no-mans land between modern cloud architectures and classical UNIX architecture. In cloud architectures, everybody is using systemd, but nobody talks about it. In the UNIX sphere, everybody is using systemd and complains about it.

I'm using VPS or dedicated servers in the cloud, mostly because these give me more control over costs and on a buck per cpu/ram/disk basis they are also generally cheaper. But this also allows me to keep up a barrier between the top 5 global tech companies and their control over my and my customers data.

So my stuff is in the cloud or on premise, but I am still living in the world of servers and as such do not have access to services modern architectures provide.

When I'm building my architecture, I want to benefit from all the good stuff kubernetes & co provide, but I have to do it the hard way, because my VPS or on-premise nodes operate on an insecure network. A cloud user can just use AWS secrets because trusting that service is no more risk than trusting the provider with everything else. In my scenario, adding a secret service does not really help me, because no secret server I can setup is more secure than what that server itself needs to be secure - secure storage of secrets.

Yes I can install and setup Vault, but a service still needs to authenticate to vault (using a secret) and vault still needs to be unsealed on boot, needing - guess what - a secret. So whatever I try to fix this, all I achieve is to add complexity instead of adding security.

So no matter what I do, I end up with a secret persistently stored in a file or a watchdog process running in some save space that monitors my services and provides them with secrets. And who should that watchdog be? I say systemd.

Now I am perfectly aware that my situation is fringe as fringe can be. People just use the cloud (so do I in many instances and I'm happy). But still I believe that while using the cloud we should still be able to do it our way, reliably. The place where systemd is used most (as part of the cloud provider infrastructure) is where my use case is irrelevant. But my place is where systemd could make a real difference.

Professionally I tend to believe that it is irresponsible not to use cloud providers to host services, if for no better reason than that it's too expensive (=difficult/insecure/fringe) to protect secrets on a Linux system. Socially or politically, I tend to believe the opposite, it's irresponsible to let cloud providers gain such a dominant position. Systemd seems to be the one party that is best equipped to restore an equilibrium in this no-mans land between people managing their own data badly and cloud providers allowing us to access to our data.

It's really very similar to what Letsencrypt did to confidentiality of data in transit. If systemd would provide a reliable solution to how to keep secrets safe (as safe as the circumstances allow for), then I would have no issue to strongly support people having their own private clouds in the cloud from which they could shout 'Halleluhjah, lluja sog I' and feel secure on their VPS.

**Comment by mutech** [comment]
> So, you want an unprivileged user to pass a secret to systemd, to hand it to a differently privileged service – all while prohibiting the originating user to read it back?

When you put it like that, it sounds pretty stupid. I want to make sure that only the entity that requires a secret can access it.

In most cases, that is a service that's supposed to run unattended as an unprivileged user and that is typically controlled by systemd. 

The originating user does not need to read back the secret, they are supposed to know it already. An attacker impersonating as the originating user however should not be allowed to read that secret.

The rationale is that once you can execute code as attacker, the service user ID is probably compromised. If however your code does not execute in the service process that systemd started (an assumption), it would still not permit the attacker to gain access of that secret. If even the service process can read the secret only once, even if the attackers code runs in that process, the secret would not be available from systemd.

> Privileged users can drop stuff in /etc/credstore/ and /run/credstore where only privileged code can read it

This is actually pretty close to what I wanted in the secret storage part, or even better, if I understand the mechanism correctly.

So if `LoadCredential` is not an absolute path, systemd searches in `/etc/credstore` et. al.  for a matching file? Any service unit can request such a secret if they know the key/path and I guess the search is performed with service permissions, so that secrets in /etc/credstore can be protected by file system permissions?

How persistent are files in /etc/credstore? If `LoadCredential` is used with an absolute path, the secret is read from that path and according to my understanding, that file has to stay there for systemd or in other words, systemd does not copy the secret to permanent storage. This has the problem I mentioned in an earlier comment, that these files are exactly what I want to get rid of. `/etc/credstore` however is fine.

Systemd uses `/var/lib/systemd/credentials.secret` to decrypt credentials stored in /etc/credstore. Does this secret get rotated?

So when deploying a service, I would encrypt the secret and as root store it in /etc/credstore.encrypted, set ownership and permissions so that only my service can access the file and systemd would take care to exposed it to my service via LoadCredential? That would survive a reboot?

That would be great. It's not quite what I wanted but more than good enough. I didn't know about all that, since all distros I'm currently working with don't have this facility, seems to be pretty new (or I'm slow).

Thanks for the great work! Really made my day (unless you tell me that no, /etc/credstore will be wiped at boot time, but nobody would be so cruel) :-)

**Comment by poettering** [comment]
> So if `LoadCredential` is not an absolute path, systemd searches in `/etc/credstore` et. al. for a matching file? Any service unit can request such a secret if they know the key/path and I guess the search is performed with service permissions, so that secrets in /etc/credstore can be protected by file system permissions?

No. systemd runs with privileges, and picks it up, and then drops it into a private ramfs instance for the target service and locks it down there. So the idea is that /etc/credstore/ can be locked down via permissions as much as possible (i.e. you can make all files 000 if you like), but systemd has CAP_DAC_OVERRIDE and thus the access mode doesn't matter for it. When it copies the creds into the per-service ramfs it will chown them to the service's own uid, and set it to 0600, so that only the service can access it.

creds are immutable (i.e. cannot be revoked, deleted, updated) as long as the service is running.

**Comment by poettering** [comment]
> How persistent are files in /etc/credstore? If `LoadCredential` is used with an absolute path, the secret is read from that path and according to my understanding, that file has to stay there for systemd or in other words, systemd does not copy the secret to permanent storage. This has the problem I mentioned in an earlier comment, that these files are exactly what I want to get rid of. `/etc/credstore` however is fine.

Where do you think it should copy it, that was any safer that /etc/credstore? I mean, they have to be kept around somewhere, and /etc/credstore (or /run/credstore) is as good as any other place, in particular if you assume people use encryption/authentication via LoadEncryptedCredential=

**Comment by mutech** [comment]
> No. systemd runs with privileges, and picks it up, and then drops it into a private ramfs instance for the target service

Ok, but how do you then isolate secrets from services? Any user can reference a secret if the know the name (key/path). So if there is no association between secret and service and FS permissions are ignored, how do you keep a service from finding secrets meant for others services? I must be missing something...

> No. systemd runs with privileges, and picks it up, and then drops it into a private ramfs instance for the target service

Yeah, that makes sense the way it's implemented.

**Comment by poettering** [comment]
> So when deploying a service, I would encrypt the secret and as root store it in /etc/credstore.encrypted, set ownership and permissions so that only my service can access the file and systemd would take care to exposed it to my service via LoadCredential? That would survive a reboot?

you'd encrypt it, put it in /etc/credstore or /etc/credstore.encrypted. YOu'd lock perms down, so that noone unpriv can read it, but systemd can read it anyway. then when a relevant service starts the relevant secret will be copied into an unswappable ramfs with restrictive perms, chowned to the service's user. the ramfs gets destroyed once the service shuts down again.

**Comment by mutech** [comment]
> [...] all protected through the TPM

That would certainly solve most of the issue, certainly the storage related part.

> then when a relevant service starts the relevant secret will be copied

How is this relation between service and secret established? Is it only the knowledge of the secret's name?

If I would use random keys that might be practically enough protection assuming that there is some limit how often a malicious service can restart with a different LoadCredential setting before something or someone stops it.

**Comment by mutech** [comment]
Btw., thanks for taking the time to discuss this in so much detail. That's really helpful!

**Comment by mutech** [comment]
> the ramfs gets destroyed once the service shuts down again

If the service could notify systemd that it's done with the ramfs, that would solve my other issue. But that's secondary anyway.

**Comment by poettering** [comment]
> Thanks for the great work! Really made my day (unless you tell me that no, /etc/credstore will be wiped at boot time, but nobody would be so cruel) :-)

we search both /etc/credstore and /run/credstore. The former is persistent, the latter a tmpfs that is hence implicitly flushed on reboot.

**Comment by poettering** [comment]
> > 
> 
> Ok, but how do you then isolate secrets from services? Any user can reference a secret if the know the name (key/path). So if there is no association between secret and service and FS permissions are ignored, how do you keep a service from finding secrets meant for others services? I must be missing something...

service unit definitions are a privileged concept, not an unprivileged one. if you are able to drop in service unit definitions, then you are god, and hence it's also fine if in it you say access any file in the whole hierarchy.

service code are an unprivileged concept. They are supposed to run unprivileged as much as possible (uid != 0, no caps, namespaces, seccomp, …), and thus cannot access any secrets in /etc/credstore, but can safely access secrets in their local ramfs instance that $CREDENTIALS_DIRECTORY points to)

**Comment by mutech** [comment]
> service unit definitions are a privileged concept, not an unprivileged one. if you are able to drop in service unit definitions, then you are god, and hence it's also fine if in it you say access any file in the whole hierarchy.

I hope I'm wrong: An unprivileged service user can create a new service unit (--user) and request a secret in there, right? So can any unprivileged user (afaik, I'm one of those dinosaurs not yet intimate with systemd). That might be configurable, but on my recent distros, this is how it works by default. If I'm right, the only protection secrets in credstores have from any user would be whether their name can be guessed or not.

If system and user credstores are completely separated, then all is well. But then the question is where these user domain secrets are stored. Then the whole loop would start at the beginning. If that user credstore is likewise isolated from other unprivileged users, and then maybe even from direct access by the service user itself, I'm happy.

**Comment by poettering** [comment]
well, on UNIX the primary and relevant security boundary is between UIDs. Thus, once "systemd --user" services are your focus, a major chunk of all protections is flushed down the toilet. We cannot do UID based controls, simply because everything in --user mode runs under the same UID anyway.

When your focus are --user service, the credentials concept there is much weaker. No ramfs for example (as mounts require privs), no chowning (as we can't dro pprivs further), nothing. No TPM stuff (since unpriv users dont get access to it)

what I wrote above is about system services (which might drop privs though, via User= or similar)

**Comment by mutech** [comment]
I don't know the architecture of systemd. My assumption was that when systemd starts a user service, that something similar to a user login happens, except that it's not a shell or desktop but a service executable that's started. Starting a session must be a privileged operation, so before control is handed over to the service, I would expect systemd to still have the privileges required to access secrets, create a ramfs, do mounts or access TPM.

When I enabled lingering mode for that service user, I did not (and still don't) expect there to be a lingering session waiting for a service activation, because that would not be enough to restore that session after reboot and if a session gets created then, there is no need to keep an idle session around. I actually don't really see a need for user sessions, considering that I can just as well start a session as system service for that user. The only reason why I know they exist is because docker-rootless uses them in their default rootless setup.

I considered user sessions to just be a redundant way to do the same thing that a system service starting a process as some user does. If there is a difference, I don't really understand the purpose of it, neither a technical necessity. But I probably just don't know enough to understand it.

So if I interpret you correctly, using system services solves my concern that the service user might get access to system secrets because a user service never has access to them. User services can only read data that is available to their UID. When systemd processes a control request, it does so as the user calling systemctl and if I enable lingering, systemd will drop all privileges before processing the service start action (f.e. on reboot). Thus no system secrets.

But two different service users still have their distinct secrets, the only issue concerning this ticket is, that secrets are actually stored in some location in the users home directory. So if I actually want them handled by systemd, I need to define a system service configured to run as that user. That's fine.

Then I just wonder why docker-rootless uses user services in first place (in their rootless-setup script). If it didn't, I would never have gone that way (defining a home directory for a service user is a strange concept for me).

So if I get that right, the intention of user services is mostly to replace old school things like `at`, letting actual human users run service-like things? Are there any actual users using this for user stuff?

But anyway, what you implemented with credstores is very much the important part of what I was looking for. I'm a bit frustrated that I have to change the whole scripting using the rootless approach, but I never liked that much anyway. Even more annoying, this will probably lead to having to replicate the whole setup of where images and stuff lives. Most annoying.

I'm closing this and again, thanks a lot for taking the time and even more so for this set of features. It's very much appreciated!
